sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala

```diff
@@ -448,7 +448,21 @@ case class Pivot(
   }
 }
 
-case class Limit(limitExpr: Expression, child: LogicalPlan) extends UnaryNode {
+/** Factory for constructing new `Limit` nodes. */
+object Limit {
+  def apply(limitExpr: Expression, child: LogicalPlan): Limit = {
+    new Limit(limitExpr, child, optimized = false)
+  }
+}
+
+/**
+ * Take the first `limitExpr` rows.
+ * @param limitExpr The number of returned rows
+ * @param child     Child operator
+ * @param optimized This node has been optimized. Note that this is only a flag marker used
+ *                  to avoid adding extra `Limit` nodes to the child operators more than once.
+ */
+case class Limit(limitExpr: Expression, child: LogicalPlan, optimized: Boolean) extends UnaryNode {
```

Comment:
I'm not a huge fan of modifying the logical plan to include extra details about what has been optimized.  This is an API that is used externally even though we don't make the same promises about it.

Can we add `def limit: Option[Expression]` or something to `LogicalPlan`?  Nodes that are 1-1 can just output the limit of their child.  The default implementation can return `None`.  You can then check this before inserting a limit in the optimizer to make sure that its not a duplicate.